syntax = "proto2";
option optimize_for = SPEED;
option cc_enable_arenas = false;

package AxisPlugs;

message BVH_VECTOR {
  required float X = 1;
  required float Y = 2;
  required float Z = 3;
}

message BVH_JOIN {
  enum Channel {
    Xposition = 0;
    Yposition = 1;
    Zposition = 2;
    Xrotation = 3;
    Yrotation = 4;
    Zrotation = 5;
  }

  required string joinName = 1;
  repeated Channel Channels = 2;
  required BVH_VECTOR Offset = 3;
  optional BVH_VECTOR EndSite = 4;
  repeated BVH_JOIN ChildJoin = 5;
}

message BVH_FRAME {
  repeated float MotionData = 1;
  optional uint32 ActorTag = 2;
  optional string ActorName = 3;
}

message BVH_HEADER {
  required BVH_JOIN RootJoin = 1;
  optional uint32 ActorTag = 2;
  optional string ActorName = 3;
}

message TAG_MAP_STRING {
  map<uint32, string> Map = 1;
}

message ERROR_MESSAGE {
  required uint32 ErrorCode = 1;
  required string ErrorMsg = 2;
}

message COMMAND {
  enum CommandType {
    Notify_Characterised = 1;
    Notify_Uncharacterised = 2;
    Notify_ActorsList = 3;
    Notify_Error = 5;
    Command_Query_BVH_Header = 6;
    Command_Query_ActorList = 7;
    Command_ZeroOutActor = 8;
  }

  required CommandType Content = 1;
  optional uint32 ActorTag = 2;
  optional TAG_MAP_STRING ActorsList = 3;
  optional ERROR_MESSAGE ErrorMsg = 4;
}

message AXIS_PLUGIN_MESSAGE {
  enum MessageType {
    COMMAND = 0;
    HEADER_DATA = 1;
    FRAME_DATA = 2;
  }

  required MessageType Message = 1;
  oneof Content {
    BVH_HEADER Header = 2;
    BVH_FRAME Frame = 3;
    COMMAND Command = 4;
  }
}